File: apps/api/src/application/services/events.service.ts
import { Injectable, BadRequestException } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import {
  CreateEventCommand,
  UpdateEventCommand,
  DeleteEventCommand,
} from '../commands';
import { GetEventsQuery } from '../queries';
import {
  CreateEventDto,
  UpdateEventDto,
  EventQueryDto,
  EventResponseDto,
} from '../dtos';
import { UsersService } from './users.service';

@Injectable()
export class EventsService {
  constructor(
    private readonly commandBus: CommandBus,
    private readonly queryBus: QueryBus,
    private readonly usersService: UsersService,
  ) {}

  async getEvents(
    userId: string,
    query: EventQueryDto,
  ): Promise<EventResponseDto[]> {
    const coupleId = await this.usersService.getUserCoupleId(userId);
    if (!coupleId) {
      return [];
    }

    const startDate = query.startDate ? new Date(query.startDate) : undefined;
    const endDate = query.endDate ? new Date(query.endDate) : undefined;

    const events = await this.queryBus.execute(
      new GetEventsQuery(userId, coupleId, startDate, endDate),
    );

    return events.map(
      (event: {
        id: string;
        coupleId: string;
        title: string;
        startTime: string;
        endTime: string;
        category: string;
        authorId: string;
        description?: string;
        createdAt: string;
        updatedAt: string;
      }) => {
        const dto = new EventResponseDto();
        dto.id = event.id;
        dto.coupleId = event.coupleId;
        dto.title = event.title;
        dto.startTime = event.startTime;
        dto.endTime = event.endTime;
        dto.category = event.category as EventResponseDto['category'];
        dto.authorId = event.authorId;
        dto.description = event.description;
        dto.createdAt = event.createdAt;
        dto.updatedAt = event.updatedAt;
        return dto;
      },
    );
  }

  async createEvent(
    userId: string,
    dto: CreateEventDto,
  ): Promise<EventResponseDto> {
    const coupleId = await this.requireCoupleId(userId);

    const event = await this.commandBus.execute(
      new CreateEventCommand(
        userId,
        coupleId,
        dto.title,
        new Date(dto.startTime),
        new Date(dto.endTime),
        dto.category,
        dto.description,
      ),
    );

    return EventResponseDto.fromEntity(event);
  }

  async updateEvent(
    userId: string,
    eventId: string,
    dto: UpdateEventDto,
  ): Promise<EventResponseDto> {
    const coupleId = await this.requireCoupleId(userId);

    const event = await this.commandBus.execute(
      new UpdateEventCommand(
        userId,
        coupleId,
        eventId,
        dto.title,
        dto.startTime ? new Date(dto.startTime) : undefined,
        dto.endTime ? new Date(dto.endTime) : undefined,
        dto.category,
        dto.description,
      ),
    );

    return EventResponseDto.fromEntity(event);
  }

  async deleteEvent(userId: string, eventId: string): Promise<void> {
    const coupleId = await this.requireCoupleId(userId);

    await this.commandBus.execute(
      new DeleteEventCommand(userId, coupleId, eventId),
    );
  }

  private async requireCoupleId(userId: string): Promise<string> {
    const coupleId = await this.usersService.getUserCoupleId(userId);
    if (!coupleId) {
      throw new BadRequestException('User is not in a couple');
    }
    return coupleId;
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/application/service/EventsService.kt
package com.couplecalendar.application.service

import com.couplecalendar.application.command.event.CreateEventCommand
import com.couplecalendar.application.command.event.CreateEventCommandHandler
import com.couplecalendar.application.command.event.DeleteEventCommand
import com.couplecalendar.application.command.event.DeleteEventCommandHandler
import com.couplecalendar.application.command.event.UpdateEventCommand
import com.couplecalendar.application.command.event.UpdateEventCommandHandler
import com.couplecalendar.application.dto.request.CreateEventRequest
import com.couplecalendar.application.dto.request.EventQueryRequest
import com.couplecalendar.application.dto.request.UpdateEventRequest
import com.couplecalendar.application.dto.response.EventResponse
import com.couplecalendar.application.query.event.GetEventsQuery
import com.couplecalendar.application.query.event.GetEventsQueryHandler
import com.couplecalendar.common.exception.BadRequestException
import com.couplecalendar.domain.aggregate.EventCategory
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.UUID

@Service
class EventsService(
    private val createEventCommandHandler: CreateEventCommandHandler,
    private val updateEventCommandHandler: UpdateEventCommandHandler,
    private val deleteEventCommandHandler: DeleteEventCommandHandler,
    private val getEventsQueryHandler: GetEventsQueryHandler,
    private val usersService: UsersService
) {

    fun getEvents(userId: UUID, queryRequest: EventQueryRequest): List<EventResponse> {
        val coupleId = requireCoupleId(userId)

        val startDate = queryRequest.startDate?.let { Instant.parse(it) }
        val endDate = queryRequest.endDate?.let { Instant.parse(it) }

        val query = GetEventsQuery(userId, coupleId, startDate, endDate)
        val results = getEventsQueryHandler.handle(query)

        return results.map { result ->
            EventResponse(
                id = result.id,
                coupleId = result.coupleId,
                title = result.title,
                startTime = result.startTime,
                endTime = result.endTime,
                category = result.category,
                authorId = result.authorId,
                description = result.description,
                location = result.location,
                createdAt = result.createdAt,
                updatedAt = result.updatedAt
            )
        }
    }

    fun createEvent(userId: UUID, request: CreateEventRequest): EventResponse {
        val coupleId = requireCoupleId(userId)

        val command = CreateEventCommand(
            userId = userId,
            coupleId = coupleId,
            title = request.title,
            startTime = Instant.parse(request.startTime),
            endTime = Instant.parse(request.endTime),
            category = EventCategory.valueOf(request.category.uppercase()),
            description = request.description,
            location = request.location
        )

        val event = createEventCommandHandler.handle(command)
        return EventResponse.fromAggregate(event)
    }

    fun updateEvent(userId: UUID, eventId: UUID, request: UpdateEventRequest): EventResponse {
        val coupleId = requireCoupleId(userId)

        val command = UpdateEventCommand(
            userId = userId,
            coupleId = coupleId,
            eventId = eventId,
            title = request.title,
            startTime = request.startTime?.let { Instant.parse(it) },
            endTime = request.endTime?.let { Instant.parse(it) },
            category = request.category?.let { EventCategory.valueOf(it.uppercase()) },
            description = request.description,
            location = request.location
        )

        val event = updateEventCommandHandler.handle(command)
        return EventResponse.fromAggregate(event)
    }

    fun deleteEvent(userId: UUID, eventId: UUID) {
        val coupleId = requireCoupleId(userId)

        val command = DeleteEventCommand(userId, coupleId, eventId)
        deleteEventCommandHandler.handle(command)
    }

    private fun requireCoupleId(userId: UUID): UUID {
        return usersService.getUserCoupleId(userId)
            ?: throw BadRequestException("User is not in a couple")
    }
}
